### DISCUSSION TITLE
rdkit on windows 10 - ImportError: DLL load failed when importing Chem

### DISCUSSION THREAD

**Discussion by YoavShamir5** [rdkit/rdkit#6268] [Q&A]
Hi, 

I created a conda env ("fresh") and installed Pytorch and then Rdkit (command: conda install -n fresh -c conda-forge rdkit) successfully .
I also installed some additional libraries to the conda env (torch-scatter, torch-cluster, e3nn, biopython; conda list below). 

Attempting to import Chem results in an ImportError:

import rdkit
from rdkit import Chem

---------------------------------------------------------------------------
ImportError                               Traceback (most recent call last)
Cell In[4], line 2
      1 import rdkit
----> 2 from rdkit import Chem

File [c:\Users\SYOAV\anaconda3\envs\fresh\lib\site-packages\rdkit\Chem\__init__.py:23](file:///C:/Users/SYOAV/anaconda3/envs/fresh/lib/site-packages/rdkit/Chem/__init__.py:23)
     21 _HasSubstructMatchStr = rdchem._HasSubstructMatchStr
     22 from rdkit.Chem.rdchem import *
---> 23 from rdkit.Chem.rdmolfiles import *
     24 from rdkit.Chem.rdmolops import *
     25 from rdkit.Chem.rdCIPLabeler import *

**ImportError: DLL load failed while importing rdmolfiles: The specified procedure could not be found.**

I am not sure what is the cause of this issue, and would appreciate your input.
I should note I am using VS Code and running the import from an ipynb file and ipykernel to use the conda env as the kernel.

Thanks,
Yoav 

PS: here is the output of conda list:

# packages in environment at C:\Users\SYOAV\anaconda3\envs\fresh:
#
# Name                    Version                   Build  Channel
asttokens                 2.2.1              pyhd8ed1ab_0    conda-forge    
backcall                  0.2.0              pyh9f0ad1d_0    conda-forge    
backports                 1.0                pyhd8ed1ab_3    conda-forge    
backports.functools_lru_cache 1.6.4              pyhd8ed1ab_0    conda-forge
biopython                 1.81            py310h8d17308_0    conda-forge    
blas                      1.0                         mkl
boost                     1.78.0          py310h220cb41_4    conda-forge    
boost-cpp                 1.78.0               h5b4e17d_0    conda-forge    
brotli                    1.0.9                hcfcfb64_8    conda-forge    
brotli-bin                1.0.9                hcfcfb64_8    conda-forge    
brotlipy                  0.7.0           py310h2bbff1b_1002
bzip2                     1.0.8                he774522_0
ca-certificates           2022.12.7            h5b45459_0    conda-forge    
cairo                     1.16.0            h60892f0_1002    conda-forge    
certifi                   2022.12.7          pyhd8ed1ab_0    conda-forge    
cffi                      1.15.1          py310h2bbff1b_3
charset-normalizer        2.0.4              pyhd3eb1b0_0
colorama                  0.4.6              pyhd8ed1ab_0    conda-forge
comm                      0.1.3              pyhd8ed1ab_0    conda-forge
contourpy                 1.0.7           py310h232114e_0    conda-forge
cryptography              39.0.1          py310h21b164f_0
cuda-cccl                 12.1.55                       0    nvidia
cuda-cudart               11.7.99                       0    nvidia
cuda-cudart-dev           11.7.99                       0    nvidia
cuda-cupti                11.7.101                      0    nvidia
cuda-libraries            11.7.1                        0    nvidia
cuda-libraries-dev        11.7.1                        0    nvidia
cuda-nvrtc                11.7.99                       0    nvidia
cuda-nvrtc-dev            11.7.99                       0    nvidia
cuda-nvtx                 11.7.91                       0    nvidia
cuda-runtime              11.7.1                        0    nvidia
cycler                    0.11.0             pyhd8ed1ab_0    conda-forge
debugpy                   1.6.6           py310h00ffb61_0    conda-forge
decorator                 5.1.1              pyhd8ed1ab_0    conda-forge
e3nn                      0.5.1              pyhd8ed1ab_0    conda-forge
executing                 1.2.0              pyhd8ed1ab_0    conda-forge
filelock                  3.9.0           py310haa95532_0
flit-core                 3.8.0           py310haa95532_0
fonttools                 4.39.3          py310h8d17308_0    conda-forge
freetype                  2.12.1               ha860e81_0
giflib                    5.2.1                h8cc25b3_3
greenlet                  2.0.2           py310h00ffb61_0    conda-forge
icc_rt                    2022.1.0             h6049295_2
icu                       64.2                 he025d50_1    conda-forge
idna                      3.4             py310haa95532_0
importlib-metadata        6.1.0              pyha770c72_0    conda-forge
importlib_metadata        6.1.0                hd8ed1ab_0    conda-forge
intel-openmp              2021.4.0          haa95532_3556
ipykernel                 6.22.0             pyh025b116_0    conda-forge
ipython                   8.12.0             pyh08f2357_0    conda-forge
jedi                      0.18.2             pyhd8ed1ab_0    conda-forge
jinja2                    3.1.2           py310haa95532_0
jpeg                      9e                   h2bbff1b_1
jupyter_client            8.1.0              pyhd8ed1ab_0    conda-forge
jupyter_core              5.3.0           py310h5588dad_0    conda-forge
kiwisolver                1.4.4           py310h232114e_1    conda-forge
lerc                      3.0                  hd77b12b_0
libbrotlicommon           1.0.9                hcfcfb64_8    conda-forge
libbrotlidec              1.0.9                hcfcfb64_8    conda-forge
libbrotlienc              1.0.9                hcfcfb64_8    conda-forge
libcublas                 11.10.3.66                    0    nvidia
libcublas-dev             11.10.3.66                    0    nvidia
libcufft                  10.7.2.124                    0    nvidia
libcufft-dev              10.7.2.124                    0    nvidia
libcurand                 10.3.2.56                     0    nvidia
libcurand-dev             10.3.2.56                     0    nvidia
libcusolver               11.4.0.1                      0    nvidia
libcusolver-dev           11.4.0.1                      0    nvidia
libcusparse               11.7.4.91                     0    nvidia
libcusparse-dev           11.7.4.91                     0    nvidia
libdeflate                1.17                 h2bbff1b_0
libffi                    3.4.2                hd77b12b_6
libnpp                    11.7.4.75                     0    nvidia
libnpp-dev                11.7.4.75                     0    nvidia
libnvjpeg                 11.8.0.2                      0    nvidia
libnvjpeg-dev             11.8.0.2                      0    nvidia
libpng                    1.6.39               h8cc25b3_0
libsodium                 1.0.18               h8d14728_1    conda-forge
libtiff                   4.5.0                h6c2663c_2
libuv                     1.44.2               h2bbff1b_0
libwebp                   1.2.4                hbc33d0d_1
libwebp-base              1.2.4                h2bbff1b_1
lz4-c                     1.9.4                h2bbff1b_0
markupsafe                2.1.1           py310h2bbff1b_0
matplotlib-base           3.7.1           py310h51140c5_0    conda-forge
matplotlib-inline         0.1.6              pyhd8ed1ab_0    conda-forge
mkl                       2021.4.0           haa95532_640
mkl-service               2.4.0           py310h2bbff1b_0
mkl_fft                   1.3.1           py310ha0764ea_0
mkl_random                1.2.2           py310h4ed8f06_0
mpmath                    1.2.1           py310haa95532_0
munkres                   1.1.4              pyh9f0ad1d_0    conda-forge
nest-asyncio              1.5.6              pyhd8ed1ab_0    conda-forge
networkx                  2.8.4           py310haa95532_1
numpy                     1.23.5          py310h60c9a35_0
numpy-base                1.23.5          py310h04254f7_0
openssl                   1.1.1t               hcfcfb64_0    conda-forge
opt-einsum                3.0.0                      py_0    conda-forge
opt_einsum_fx             0.1.4              pyhd8ed1ab_0    conda-forge
packaging                 23.0               pyhd8ed1ab_0    conda-forge
pandas                    1.5.3           py310h1c4a608_1    conda-forge
parso                     0.8.3              pyhd8ed1ab_0    conda-forge
pickleshare               0.7.5                   py_1003    conda-forge
pillow                    9.4.0           py310hd77b12b_0
pip                       23.0.1          py310haa95532_0
pixman                    0.38.0            hfa6e2cd_1003    conda-forge
platformdirs              2.5.2           py310haa95532_0
prompt-toolkit            3.0.38             pyha770c72_0    conda-forge
prompt_toolkit            3.0.38               hd8ed1ab_0    conda-forge
psutil                    5.9.4           py310h8d17308_0    conda-forge
pure_eval                 0.2.2              pyhd8ed1ab_0    conda-forge
pycairo                   1.23.0          py310h42c1a3e_0    conda-forge
pycparser                 2.21               pyhd3eb1b0_0
pygments                  2.14.0             pyhd8ed1ab_0    conda-forge
pyopenssl                 23.0.0          py310haa95532_0
pyparsing                 3.0.9              pyhd8ed1ab_0    conda-forge
pysocks                   1.7.1           py310haa95532_0
python                    3.10.10              h966fe2a_2
python-dateutil           2.8.2              pyhd8ed1ab_0    conda-forge
python_abi                3.10                    2_cp310    conda-forge
pytorch                   2.0.0           py3.10_cuda11.7_cudnn8_0    pytorch
pytorch-cuda              11.7                 h16d0643_3    pytorch
pytorch-mutex             1.0                        cuda    pytorch
pytz                      2023.3             pyhd8ed1ab_0    conda-forge
pywin32                   304             py310h00ffb61_2    conda-forge
pyzmq                     25.0.2          py310hcd737a0_0    conda-forge
rdkit                     2022.09.5       py310h5222231_0    conda-forge
reportlab                 3.6.12          py310hf830693_2    conda-forge
requests                  2.28.1          py310haa95532_1
scipy                     1.10.1                   pypi_0    pypi
setuptools                65.6.3          py310haa95532_0
six                       1.16.0             pyhd3eb1b0_1
sqlalchemy                1.4.46          py310h8d17308_0    conda-forge
sqlite                    3.41.1               h2bbff1b_0
stack_data                0.6.2              pyhd8ed1ab_0    conda-forge
sympy                     1.11.1          py310haa95532_0
tk                        8.6.12               h2bbff1b_0
torch-cluster             1.6.1+pt113cu117          pypi_0    pypi
torch-scatter             2.1.1+pt113cu117          pypi_0    pypi
torchaudio                2.0.0                    pypi_0    pypi
torchvision               0.15.0                   pypi_0    pypi
tornado                   6.2             py310h8d17308_1    conda-forge
traitlets                 5.9.0              pyhd8ed1ab_0    conda-forge
typing_extensions         4.4.0           py310haa95532_0
tzdata                    2022g                h04d1e81_0
ucrt                      10.0.22621.0         h57928b3_0    conda-forge
unicodedata2              15.0.0          py310h8d17308_0    conda-forge
urllib3                   1.26.15         py310haa95532_0
vc                        14.2                 h21ff451_1
vs2015_runtime            14.34.31931         h4c5c07a_10    conda-forge
wcwidth                   0.2.6              pyhd8ed1ab_0    conda-forge
wheel                     0.38.4          py310haa95532_0
win_inet_pton             1.1.0           py310haa95532_0
wincertstore              0.2             py310haa95532_2
xz                        5.2.10               h8cc25b3_1
zeromq                    4.3.4                h0e60522_1    conda-forge
zipp                      3.15.0             pyhd8ed1ab_0    conda-forge
zlib                      1.2.13               h8cc25b3_0
zstd                      1.5.4                hd43e919_0

**Comment by greglandrum** [comment]
@YoavShamir5 what happens if you try 'python -c "from rdkit import Chem"' from a command.com (or bash) prompt where you've activated the `fresh` environment?

**Reply by YoavShamir5** [reply]
Hello Greg,

Thanks for the quick reply, running this from the Anaconda prompt seems to result in the same error:

(base) C:\Users\SYOAV>conda activate fresh

(fresh) C:\Users\SYOAV>
(fresh) C:\Users\SYOAV>python -c "from rdkit import Chem"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
  File "C:\Users\SYOAV\anaconda3\envs\fresh\lib\site-packages\rdkit\Chem\__init__.py", line 23, in <module>
    from rdkit.Chem.rdmolfiles import *
ImportError: DLL load failed while importing rdmolfiles: The specified procedure could not be found.

**Reply by greglandrum** [reply]
Something odd is going on. To remove as many variables as possible, could you please try creating an environment which only has the RDKit installed, activate that, and then run the command above?

If that works, then something else you have installed in the `fresh` environment somehow broke it in some way, and we'll need to take a look at exactly which commands you ran (and in what order) to create the environment.

**Reply by YoavShamir5** [reply]
1) I created a new conda env (called "war"), and installed Rdkit using the same command. this worked - "from rdkit import Chem" did not result in an error.
2) An attempt to install pytorch (conda install -n war pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia) in the same env results in a dependencies issue:

(war) C:\Users\SYOAV>conda install -n war pytorch torchvision torchaudio pytorch-cuda=11.7 -c pytorch -c nvidia
Collecting package metadata (current_repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: failed with repodata from current_repodata.json, will retry with next repodata source.
Collecting package metadata (repodata.json): done
Solving environment: failed with initial frozen solve. Retrying with flexible solve.
Solving environment: - 
Found conflicts! Looking for incompatible packages.   
This can take several minutes.  Press CTRL-C to abort.
Examining pytorch-cuda=11.7:  86%|██████████████████████████████████████████████████████████████████████████████████████████████████████████████████                   | 6/7 [00:11<00:01,  1.72s/it]|
failed

UnsatisfiableError: The following specifications were found
to be incompatible with the existing python installation in your environment:

Specifications:

  - torchaudio -> python[version='>=3.10,<3.11.0a0|>=3.9,<3.10.0a0|>=3.8,<3.9.0a0|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0']
  - torchvision -> python[version='>=3.10,<3.11.0a0|>=3.8,<3.9.0a0|>=3.9,<3.10.0a0|>=3.7,<3.8.0a0|>=3.6,<3.7.0a0|>=3.5,<3.6.0a0']

Your python: python=3.11

If python is on the left-most side of the chain, that's the version you've asked for.
When python appears to the right, that indicates that the thing on the left is somehow
not available for the python version you are constrained to. Note that conda will not
change your python version to a different minor version unless you explicitly specify
that.

The following specifications were found to be incompatible with each other:

Output in format: Requested package -> Available versions

Package typing conflicts for:
torchvision -> typing_extensions -> typing[version='>=3.7.4']
pytorch -> typing_extensions -> typing[version='>=3.7.4']

Package pytorch-cuda conflicts for:
torchvision -> pytorch-cuda[version='11.6.*|11.7.*|11.8.*']
torchvision -> pytorch==2.0.0 -> pytorch-cuda[version='>=11.6,<11.7|>=11.7,<11.8|>=11.8,<11.9']
torchaudio -> pytorch-cuda[version='11.6.*|11.7.*|11.8.*']
torchaudio -> pytorch==2.0.0 -> pytorch-cuda[version='>=11.6,<11.7|>=11.7,<11.8|>=11.8,<11.9']

Package requests conflicts for:
torchvision -> requests
python=3.11 -> pip -> requests

Package pytorch conflicts for:
torchvision -> pytorch[version='1.10.0|1.10.1|1.10.2|1.11.0|1.12.0|1.12.1|1.13.0|1.13.1|2.0.0|1.9.1|1.9.0|1.8.1|1.8.0|1.7.1|1.7.0|1.6.0|1.5.1']
torchaudio -> pytorch[version='1.10.0|1.10.1|1.10.2|1.11.0|1.12.0|1.12.1|1.13.0|1.13.1|2.0.0|1.9.1|1.9.0|1.8.1|1.8.0|1.7.1|1.7.0|1.6.0']

Package libzlib conflicts for:
python=3.11 -> libzlib[version='>=1.2.13,<1.3.0a0']
python=3.11 -> zlib[version='>=1.2.13,<1.3.0a0'] -> libzlib==1.2.13=hcfcfb64_4

Package setuptools conflicts for:
pytorch -> jinja2 -> setuptools
python=3.11 -> pip -> setuptools
torchvision -> setuptools

3) I should note - in "fresh" env, I was able to install first Pytorch and then Rdkit using the same commands.

Do you have an idea what could be a solution to this? If not I'll check with the Pytorch forums.

Thanks again,
Yoav

**Reply by greglandrum** [reply]
I believe that the error message is telling you that pytorch is only available for python versions <3.11, but you have 3.11 installed.
In your previous example you had python 3.10 installed, so you might try recreating your `war` environment with: `conda create -c conda-forge python=3.10 rdkit`, make sure that works (with the command above), then try installing pytorch and see what happens.

**Comment by BuiQuocDung1991** [comment]
I have the same problem when i make new env environment with below codes:
```
conda create -n env python anaconda -y
conda install -c conda-forge rdkit -y 
```
The version of python is 3.11
When i run jupiter in VS code, this error appeared: 
**ImportError: DLL load failed when importing Chem**
Then i attempt to reinstall with below code (of course i removed the previous env environment)
`conda create -c conda-forge -n env rdkit -y`
The error disappeared with python 3.10. Maybe, the issue is in the version of python. I have no idea exactly what happened but i guess.

**Comment by tim-the-wizard** [comment]
Just adding this here in case it helps save someone some time.

After creating new conda environments, Jupyter notebooks (in VS Code) would not display molecules, e.g.
```
co2 = Chem.MolFromSmiles('O=C=O')
co2
```
should draw the molecule below.

When attempting to use the Draw module, the import failed:
```
from rdkit.Chem import Draw
ImportError: DLL load failed while importing rdMolDraw2D: The specified module could not be found.
```
However, an old environment still worked. Creating a new environment with the same Python and RDKit versions did not work.

Overriding the default conda channels rectified the issue, e.g.
`conda create -n rdkit-test --override-channels -c conda-forge python=3.12 rdkit`

Presumably incorrect dependencies were selected across mixed channels (default vs conda-forge).
